src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
-		umb_available=false;
-		if (!section->Get_bool("xms")) return;
-		Bitu i;
-		BIOS_ZeroExtendedSize(true);
-		DOS_AddMultiplexHandler(multiplex_xms);
-
-		/* place hookable callback in writable memory area */
-		xms_callback=RealMake(DOS_GetMemory(0x1)-1,0x10);
-		callbackhandler.Install(&XMS_Handler,CB_HOOKABLE,Real2Phys(xms_callback),"XMS Handler");
-		// pseudocode for CB_HOOKABLE:
-		//	jump near skip
-		//	nop,nop,nop
-		//	label skip:
-		//	callback XMS_Handler
-		//	retf
-
-		for (i=0;i<XMS_HANDLES;i++) {
-			xms_handles[i].free=true;
-			xms_handles[i].mem=-1;
-			xms_handles[i].size=0;
-			xms_handles[i].locked=0;
-		}
-		/* Disable the 0 handle */
-		xms_handles[0].free	= false;
-
-		/* Set up UMB chain */
-		umb_available=section->Get_bool("umb");
-		bool ems_available = GetEMSType(section)>0;
-		DOS_BuildUMBChain(section->Get_bool("umb"),ems_available);
-	}
-
-	~XMS(){
-		Section_prop * section = static_cast<Section_prop *>(m_configuration);
-		/* Remove upper memory information */
-		dos_infoblock.SetStartOfUMBChain(0xffff);
-		if (umb_available) {
-			dos_infoblock.SetUMBChainState(0);
-			umb_available=false;
-		}
+	XMS(Section* configuration);
+	~XMS() override;
+};
 
-		if (!section->Get_bool("xms")) return;
-		/* Undo biosclearing */
-		BIOS_ZeroExtendedSize(false);
+XMS::XMS(Section* configuration) : Module_base(configuration), callbackhandler{}
+{
+	Section_prop* section = static_cast<Section_prop*>(configuration);
 
-		/* Remove Multiplex */
-		DOS_DelMultiplexHandler(multiplex_xms);
+	umb.is_available       = false;
+	a20.enable_global      = false;
+	a20.enable_local_count = 0;
 
-		/* Free used memory while skipping the 0 handle */
-		for (Bitu i = 1;i<XMS_HANDLES;i++) 
-			if(!xms_handles[i].free) XMS_FreeMemory(i);
+	if (!section->Get_bool("xms")) {
+		return;
 	}
 
-};
-static XMS* test;
+	// NTS: Disable XMS emulation if CPU type is less than a 286, because
+	// extended memory did not exist until the CPU had enough address lines
+	// to read past the 1MB mark.
+	//
+	// The other reason we do this is that there is plenty of software that
+	// assumes 286+ instructions if they detect XMS services, including but
+	// not limited to:
+	//
+	//      MSD.EXE Microsoft Diagnostics
+	//      Microsoft Windows 3.0
+	//
+	// Not emulating XMS for 8086/80186 emulation prevents the software
+	// from crashing.
+
+	if (CPU_ArchitectureType < CPU_ARCHTYPE_286) {
+		LOG_WARNING("XMS: CPU 80186 or lower lacks address lines needed for XMS, disabling");
+		return;
+	}
+
+	xms.is_available = true;
+	// TODO: read HMA configuration
+
+	BIOS_ZeroExtendedSize(true);
+	DOS_AddMultiplexHandler(xms_multiplex);
+
+	// Place hookable callback in writable memory area
+	xms.callback = RealMake(static_cast<uint16_t>(DOS_GetMemory(0x1) - 1),
+	                        0x10);
+	callbackhandler.Install(&XMS_Handler,
+	                        CB_HOOKABLE,
+	                        Real2Phys(xms.callback),
```

Comment:
Oh man... how about RealMemToPhysical? Or lower snake case if not exported.

Reply:
Renamed to `RealToPhysical`, for consistency with other function.

Note: this was mostly a search&replace commit, with some small exceptions no other cleanups were done.